src/renderer/store.ts

```diff
@@ -1,19 +1,284 @@
-import { createStore, combineReducers } from 'redux';
-import * as LocalRenderer from '@getflywheel/local/renderer';
-import { IPC_EVENTS } from '../constants';
-import { Preferences } from '../types';
-import { preferencesReducer } from './reducers'
-
-
-const makeStore = async () => {
-	const preferences: Preferences = await LocalRenderer.ipcAsync(
-		IPC_EVENTS.READ_PREFERENCES_FROM_DISK
-	);
-
-	return createStore(
-		combineReducers({ preferences: preferencesReducer }),
-		{ preferences },
-	);
+import { useSelector, TypedUseSelectorHook } from 'react-redux';
+import {
+	configureStore,
+	createSlice,
+	createSelector,
+	PayloadAction,
+} from '@reduxjs/toolkit';
+
+import {
+	Preferences,
+	CachedImageDataBySiteID,
+	SiteImageData,
+	OptimizerStatus,
+	ImageData,
+	FileStatus,
+} from '../types';
+
+interface SiteActionPayload {
+	siteID: string;
+	siteImageData: SiteImageData;
+	error: Error;
+}
+
+function mergeSiteState(state, payload: Partial<SiteActionPayload>, newState?: Partial<SiteImageData>) {
+	const { siteID, siteImageData } = payload;
+
+	state[siteID] = {
+		...state[siteID],
+		...siteImageData,
+		...newState,
+	};
+
+	return state;
+}
+
+function generateInitialSiteState(): Partial<SiteImageData> {
+	return {
+		scanInProgress: false,
+		selectAllFilesValue: true,
+		optimizationStatus: OptimizerStatus.BEFORE,
+		compressionListCompletionPercentage: 0,
+	};
+}
+
+const preferencesSlice = createSlice({
+	name: 'preferences',
+	initialState: {} as Preferences,
+	reducers: {
+		hydratePreferences: (_, action: PayloadAction<Preferences>) => {
+			return action.payload;
+		},
+		stripMetaData: (state, action: PayloadAction<boolean>) => {
+			state.stripMetaData = action.payload;
+			return state;
+		},
+	},
+});
+
+const sitesSlice = createSlice({
+	name: 'sites',
+	initialState: {} as CachedImageDataBySiteID,
+	reducers: {
+		hydrateSites: (_, action: PayloadAction<CachedImageDataBySiteID>) => {
+			const initialState = Object.entries(action.payload).reduce((acc, [id, siteImageData]) => {
+				acc[id] = {
+					...siteImageData,
+					...generateInitialSiteState(),
+				};
+
+				Object.values(acc[id].imageData).forEach((d) => d.isChecked = true);
+
+				return acc;
+			}, {} as CachedImageDataBySiteID);
+
+			return initialState;
+		},
+		addSite: (state, action: PayloadAction<{ siteID: string }>) => {
+			const { siteID } = action.payload;
+
+			state[siteID] = {
```

Comment:
I don't think we should directly modify the existing state throughout these reducers. Though I need to read up on Redux Toolkit so maybe there's something I'm missing.

Reply:
That's what I thought too, but since this uses [Immer](https://immerjs.github.io/immer/docs/introduction) under the hood, `state` is something called a `WritableDraft`. This object is intended to be mutated as it represents the "next state". Eventually, this gets written as a current state, which is immutable. Apparently the idea behind this pattern is that you only ever have two copies of state at once which can help keep the store's memory footprint smaller.

Reply:
Got you, sounds good.

Reply:
RTK uses `Immer` behind the scenes, so what Matt is modifying here is a draft of the state that is automatically merged with the existing state upon the function completing. Definitely not obvious on the surface, however, as it looks like it's mutating it. After getting used to it, it's really super cool and allows for things like this (and intelligently merges):
```
const todosReducer = createReducer(initialState, (builder) => {
  builder.addCase(addTodo, (state, action) => {
    state.push(action.payload)
    console.log(current(state))
  })
})
```

More [here](https://redux-toolkit.js.org/api/other-exports#current).